ClLike/cl_like/cl_like.py

```diff
@@ -276,222 +279,126 @@ def get_cl_data_sacc(self):
+        state['derived']['dchi2_marg'] = dchi2
+        # - Number of function evaluations
+        state['derived']['nfev'] = p.nfev
+
+    def _get_chi2(self, **pars):
+        bias_names, bias_info = self.provider.get_bias_info()
+        assert bias_names == self.bias_names
+        global_bias = {name: 1 for name in self.bin_properties.keys()}
+
+        def chi2_f(bias):
+            t, g = self.provider.get_cl_with_bias(bias, global_bias)
             r = t - self.data_vec
             ic_r = np.dot(self.inv_cov, r)
             chi2 = np.dot(r, ic_r) # (ndata) , (ndata, ndata) , (ndata)
-            g = self._model_deriv(cld, bias)
+
             dchi2 = 2*np.dot(ic_r, g) # (ndata, ndata) , (ndata) , (ndata, nbias)
             # Bias prior
             rb = bias - self.bias_pr_mean
             chi2 += np.sum(rb**2*self.bias_pr_isigma2)
             dchi2 += 2*rb*self.bias_pr_isigma2
             return chi2, dchi2
 
-        p = minimize(chi2, self.bias0, method='Newton-CG', jac=True,
-                     hess=lambda b: self.hessian_chi2(b, cld))
-        H = self.hessian_chi2(p.x, cld,
-                              include_DF=self.bias_fisher_deriv2)
-
-        return p.fun, 0.5*H, p
-
-    def get_can_provide_params(self):
-        return self.bias_names + ['nfev', 'dchi2_marg']
-
-    def calculate(self, state, want_derived=True, **pars):
-        # Calculate chi2
-        chi2, F, p = self._get_BF_chi2_and_F(**pars)
-
-        # Update starting point
-        if self.bias_update_every or (not self.updated_bias0):
-            self.bias0 = p.x.copy()
+        def hessian_chi2_f(bias, include_DF=False):
+            t, g = self.provider.get_cl_with_bias(bias, global_bias)
+            ic_g = np.dot(self.inv_cov, g) # (ndata, ndata) , (ndata, nbias)
+            ddchi2 = 2*np.sum(g[:, None, :]*ic_g[:, :, None], axis=0) # (ndata, _, nbias) , (ndata, nbias, _) -> (nbias, nbias)
+            # Bias prior
+            ddchi2 += 2*np.diag(self.bias_pr_isigma2)
+            if include_DF:
+                ddt = self.provider.get_cl_theory_dderiv(bias, global_bias)
+                r = t-self.data_vec
+                ic_r = np.dot(self.inv_cov, r)
+                ddchi2 += 2*np.sum(ic_r[:, None, None]*ddt, axis=0) # (ndata), (ndata, nbias, nbias)
+            return ddchi2
+
+        p = minimize(chi2_f, self.bias0, method='Newton-CG', jac=True,
+                     hess=lambda b: hessian_chi2_f(b))
+        chi2 = p.fun
+
+        H = hessian_chi2_f(p.x, include_DF=self.bias_fisher_deriv2)
+        ev = np.linalg.eigvals(0.5*H)
+        if np.any(ev <= 0):  # Use positive-definite version if needed
+            H = hessian_chi2_f(p.x, include_DF=False)
+            ev = np.linalg.eigvals(0.5*H)
+        dchi2 = np.sum(np.log(ev))
+        if np.isnan(dchi2):
+            print(dchi2)
+            print(ev)
+            print(H)
+            exit(1)
+
+        return chi2, dchi2, p
+
+    def get_cl_theory_sacc(self,bias,global_bias):
```

Comment:
i would not overwrite the default one. I think the default one will give you the Cells for the biases that maximize the likelihood, which I think it is interesting to have.